lib/concurrent/synchronization/java_pure_object.rb

```diff
@@ -32,6 +33,10 @@ def ns_signal
           JRuby.reference0(self).notify
           self
         end
+
+        def ensure_ivar_visibility!
+          # relying on undocumented behavior of JRuby, ivar access is volatile
```

Comment:
@pitr-ch sorry, you are 100% right.

 * reads are plain Java field and array accesses,
 * writes are volatile read + volatile write.

Therefore for example JVM is allowed to "break" the following code:
```ruby
# EXAMPLE n. 1
class Foo
  def initialize
    @all_systems_go = false
  end
  
  def go!
    @all_systems_go = true
  end
  
  def await_the_go_ahead
    nil until @all_systems_go
    puts "We're good to go!"
  end
end

foo = Foo.new
Thread.new { foo.go! }
foo.await_the_go_ahead # this is allowed to spin infinitely
```

The loop in `nil until @all_systems_go` is allowed to cache and never re-read the value of `@all_systems_go`.

However, if used correctly, the __current implementation__ (so this in no way future proof and is not an official JRuby position) does give us "i-var proof" thread safe construction of objects:

```ruby
# EXAMPLE n. 2
class Foo
  attr_reader :val
  
  def initialize
    @val = :VALUE
  end
end

class Bar
  def setup_foo
    @foo = Foo.new
  end
  
  def foo_val
    nil until foo = @foo
    @foo.val
  end
end

bar = Bar.new
Thread.new { bar.setup_foo }
# might spin infinitely (see example n. 1), however if
# a value is produced it will be :VALUE and never `nil`
# or an exception etc.
bar.foo_val
```

The above example's `bar.foo_val` is allowed to go into an infinite loop (see example n. 1), but if it does return, the return value __will be__ `:VALUE` and not a `nil` and not an exception. This is because of `volatile` semantics combined with JMM's ban on out-of-thin-air values. Note that this applies only to i-vars, so if `Foo`'s constructor was something like this:

```ruby
class Foo
  def initialize # broken
    @val = []
    @val << :VALUE
  end
  
  def good_initialize # not broken
    val = []
    val << :VALUE
    @val = val
  end
end
```

then there are no guarantees about the state of `@val` array for the other thread in `Foo#foo_val`.